app/src/main/java/com/github/sdpteam15/polyevents/database/room/LocalDatabase.kt

```diff
@@ -0,0 +1,92 @@
+abstract class LocalDatabase: RoomDatabase() {
+    abstract fun eventDao(): EventDao
+
+    companion object {
+        @Volatile
+        private var INSTANCE: LocalDatabase? = null
+
+        fun getDatabase(
+                context: Context,
+                scope: CoroutineScope
+        ): LocalDatabase {
+            // if the INSTANCE is not null, then return it,
+            // if it is, then create the database
+            return INSTANCE ?: synchronized(this) {
+                val instance = Room.databaseBuilder(
+                        context.applicationContext,
+                        LocalDatabase::class.java,
+                        "polyevents_database"
+                )
+                        // Wipes and rebuilds instead of migrating if no Migration object.
+                        // Migration is not part of this codelab.
+                        .fallbackToDestructiveMigration()
+                        .addCallback(WordDatabaseCallback(scope))
+                        .build()
+                INSTANCE = instance
+                // return instance
+                instance
+            }
+        }
+
+        /**
+         * To delete all content and repopulate the database whenever the app is created,
+         * we need to create a RoomDatabase.Callback and override onCreate().
+         * Because you cannot do Room database operations on the UI thread,
+         * onCreate() launches a coroutine on the IO Dispatcher.
+         * (ref: https://developer.android.com/codelabs/android-room-with-a-view-kotlin#13)
+         *
+         * Note: Populating the database isn't related to a UI lifecycle, therefore you shouldn't
+         * use a CoroutineScope like viewModelScope.  It's related to the app's lifecycle. Therefore
+         * use application scope defined in Polyevents Application
+         */
+        private class WordDatabaseCallback(
+                private val scope: CoroutineScope
+        ) : RoomDatabase.Callback() {
+            /**
+             * Override the onCreate method to populate the database.
+             */
+            override fun onCreate(db: SupportSQLiteDatabase) {
+                super.onCreate(db)
+                // If you want to keep the data through app restarts,
+                // comment out the following line.
+                INSTANCE?.let { database ->
+                    scope.launch(Dispatchers.IO) {
+                        populateDatabase(database.eventDao())
+                    }
+                }
+            }
+        }
+
+        /**
+         * Populate the database in a new coroutine.
+         * If you want to start with more words, just add them.
+         */
+        suspend fun populateDatabase(eventDao: EventDao) {
+            // Start the app with a clean database every time.
+            // Not needed if you only populate on creation.
+            eventDao.deleteAll()
+
+            // TODO: populate the local database with that of the remote
```

Comment:
To populate the database of events.
But `onCreate()` of `WordDatabaseCallback` is never called

```suggestion
                        populateDatabase(database.eventDao(), scope)
                    }
                }
            }
        }

        /**
         * Populate the database in a new coroutine.
         * If you want to start with more words, just add them.
         */
        suspend fun populateDatabase(eventDao: EventDao, scope: CoroutineScope) {
            // Start the app with a clean database every time.
            // Not needed if you only populate on creation.
            val ids = eventDao.getAll().map { it.eventId }
            com.github.sdpteam15.polyevents.database.Database.currentDatabase.getListEntity(
                ObservableList<Event>().observeOnce {
                    scope.launch(Dispatchers.IO) {
                        eventDao.deleteAll()
                        for (e in it.value)
                            eventDao.insert(EventLocal.fromEvent(e))
                    }
                    Unit
                }.then,
                ids,
                null,
                DatabaseConstant.CollectionConstant.EVENT_COLLECTION
            )
            // TODO: populate the local database with that of the remote
```

Reply:
Thanks will commit. This part is still to be implemented. I think this is called during app installation.